conans/test/util/detect_test.py

```diff
@@ -1,15 +1,36 @@
+import platform
 import unittest
-from conans.test.utils.tools import TestBufferConanOutput
+
+from conans import tools
 from conans.client.conf.detect import detect_defaults_settings
-import platform
+from conans.test.utils.tools import TestBufferConanOutput
 
 
 class DetectTest(unittest.TestCase):
 
-    def detect_test(self):
+    platform_default_compilers = {
+        "Linux": "gcc",
+        "Darwin": "apple-clang",
+        "Windows": "Visual Studio"
+    }
+
+    def detect_default_compilers_test(self):
         output = TestBufferConanOutput()
         result = detect_defaults_settings(output)
-        if platform.system() == "Linux":
-            for (name, value) in result:
-                if name == "compiler":
-                    self.assertEquals(value, "gcc")
+        # result is a list of tuples (name, value) so converting it to dict
+        result = dict(result)
+        platform_compiler = self.platform_default_compilers.get(platform.system(), None)
+        if platform_compiler is not None:
+            self.assertEquals(result.get("compiler", None), platform_compiler)
+
+    @unittest.skipIf(platform.system() != "Darwin", "Test only running in Mac OS X")
+    def detect_default_in_mac_os_using_gcc_as_default_test(self):
+        # See: https://github.com/conan-io/conan/issues/2231
+        output = TestBufferConanOutput()
+        with tools.environment_append({"CC": "gcc"}):
```

Comment:
This test is nice, but is not testing the reported issue. Which is gcc being actually a symlink to clang. It might be difficult to fully automate, so I am fine if tested manually.

Reply:
Indeed, this test is only executed in MacOS x (Darwin) where gcc command is clang instead clang which was the bug report for this issue.

See output:
```
17:06 $ gcc --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 9.0.0 (clang-900.0.38)
Target: x86_64-apple-darwin17.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```
and the test is checking that detected compiler using gcc should be ``apple-clang`` https://github.com/conan-io/conan/pull/2487/files#diff-32d9d50eb3e70bdef7cfb86080926414R13

Reply:
It is not to make it more evident. It could happen that the CI machine could have *gcc* really installed in it, and the test would pass without checking for clang output, so it wouldn't test at all the reported bug fix. It is necessary to explicitly check for "apple-clang" in the output

Reply:
@memsharded I don't understand what's the problem? @lasote do you know what are we talking about? This test is only executed in Mac OS X and the CI machine for Mac OS X has fake ``gcc`` as ``apple-clang``, but if you don't want to rely on this assumption maybe it's better to remove this end2end test.

Reply:
Just check the output to confirm that the assumption is true. Because if the assumption change, it is better to know it with a fail test than being happy with a green build and a useless test.

Reply:
Ok, I got it, thanks for the explanation. I wasn't getting it, sorry guys!!!